espnet/nets/time_sync_beam_search.py

```diff
@@ -0,0 +1,228 @@
+
+        if self.decoder is not None:
+            init_decoder_state = self.decoder.init_state(enc_output)
+            decoder_scores, decoder_state = self.decoder.score(
+                self.sos_th, init_decoder_state, enc_output
+            )
+            self.attn_cache[(self.sos,)] = CacheItem(
+                state=decoder_state,
+                scores=decoder_scores,
+                log_sum=0.0,
+            )
+        if self.lm is not None:
+            init_lm_state = self.lm.init_state(enc_output)
+            lm_scores, lm_state = self.lm.score(self.sos_th, init_lm_state, enc_output)
+            self.lm_cache[(self.sos,)] = CacheItem(
+                state=lm_state,
+                scores=lm_scores,
+                log_sum=0.0,
+            )
+
+    def cached_score(self, h: Tuple[int], cache: dict, scorer: ScorerInterface) -> Any:
+        root = h[:-1]  # prefix
+        if root in cache:
+            root_scores = cache[root].scores
+            root_state = cache[root].state
+            root_log_sum = cache[root].log_sum
+        else:  # run decoder fwd one step and update cache
+            root_root = root[:-1]
+            root_root_state = cache[root_root].state
+            root_scores, root_state = scorer.score(
+                torch.tensor(root).long(), root_root_state, self.enc_output
+            )
+            root_log_sum = cache[root_root].log_sum + float(
+                cache[root_root].scores[root[-1]]
+            )
+            cache[root] = CacheItem(
+                state=root_state, scores=root_scores, log_sum=root_log_sum
+            )
+        cand_score = float(root_scores[h[-1]])
+        score = root_log_sum + cand_score
+
+        return score
+
+    def joint_score(self, hyps: Any, ctc_score_dp: Any) -> Any:
+        scores = dict()
+        for h in hyps:
+            score = self.ctc_weight * np.logaddexp(*ctc_score_dp[h])  # ctc score
+            if len(h) > 1 and self.decoder_weight > 0 and self.decoder is not None:
+                score += (
+                    self.cached_score(h, self.attn_cache, self.decoder)
+                    * self.decoder_weight
+                )  # attn score
+            if len(h) > 1 and self.lm_weight > 0 and self.lm is not None:
+                score += (
+                    self.cached_score(h, self.lm_cache, self.lm) * self.lm_weight
+                )  # lm score
+            score += self.penalty * (len(h) - 1)  # penalty score
+            scores[h] = score
+        return scores
+
+    def time_step(self, p_ctc: Any, ctc_score_dp: Any, hyps: Any) -> Any:
+        pre_beam_threshold = np.sort(p_ctc)[-self.pre_beam_size]
+        cands = set(np.where(p_ctc >= pre_beam_threshold)[0])
+        if len(cands) == 0:
+            cands = {np.argmax(p_ctc)}
+        new_hyps = set()
+        ctc_score_dp_next = defaultdict(
+            lambda: (float("-inf"), float("-inf"))
+        )  # (p_nb, p_b)
+        tmp = []
+        for l in hyps:
```

Comment:
I recommend you add some `logging.debug` functions of how the hypothesis grows in each time step.